File: solutions/179.Largest_Number/AC_sort_nlogn.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_sort_nlogn.cpp
*  Create Date: 2015-02-26 10:13:18
*  Descripton:  sort + cmp
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
private:
	static bool cmp(string a, string b) {
		return (a + b) > (b + a);
	}

public:
    string largestNumber(vector<int> &num) {
		vector<string> str;
		for (auto &i : num)
			str.push_back(to_string(i));

		sort(str.begin(), str.end(), cmp);

		string ans;
		if (str[0] == "0")
			return "0";
		for (auto &i : str)
			ans += i;

		return ans;
    }
};

int main() {
	int n, t;
	Solution s;
	cin >> n;
	vector<int> num(n);
	for (auto &i : num)
		cin >> i;
	cout << s.largestNumber(num) << endl;
	return 0;
}

File: solutions/179.Largest_Number/AC_sort_nlogn.java
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_sort_nlogn.java
*  Create Date: 2015-02-26 10:40:55
*  Descripton:   
*/

import java.util.Comparator;
import java.util.Scanner;

public class Solution {

    public String largestNumber(int[] num) {
        String[] str = new String[num.length];
        for (int i = 0; i < num.length; ++i) {
            str[i] = num[i] + "";
        }

        Comparator<String> comp = new Comparator<String>() {
            @Override
            public int compare(String o1, String o2) {
                return (o2 + o1).compareTo(o1 + o2);    // reverse
            }
        };

        Arrays.sort(str, comp);
        if (str[0].equals("0")) {
            return "0";
        }

        String ans = new String();
        for (String s : str)
            ans += s;

        return ans;
    }

    // debug
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        Solution s = new Solution();
        int[] input = {0, 0, 0};
        System.out.println(s.largestNumber(input));
    }
}
